### PR TITLE
Tiered rejit work items

### PR DESCRIPTION
**Pull request by briansull**
1. Use info.compFullName as the input to create the JIT's MethodHash
2. Update JitOrder to print out MethodHash and PerfScore
3. Change eeGetMethodFullName to expand class and struct names for the argument types and the return type
4. fixed issue where bad edge weight were set in fgFoldConditional
- Made flEdgeWeightMin and flEdgeWeightMax private fields
- Added new method setNewEdgeWeights
- Added support method eeGetArgClass

### CODE REVIEW

File: src/jit/block.h [hunk 1]
```diff
@@ -1274,13 +1274,23 @@ struct BasicBlockList
...
+    BasicBlock* flBlock;    // The BasicBlock of interest.
+    unsigned    flDupCount; // The count of duplicate "edges" (use only for switch stmts)
 
+private:
     BasicBlock::weight_t flEdgeWeightMin;
     BasicBlock::weight_t flEdgeWeightMax;
 
-    unsigned flDupCount; // The count of duplicate "edges" (use only for switch stmts)
+public:
+    BasicBlock::weight_t edgeWeightMin()
```
Comment (BruceForstall) [inline src/jit/block.h]: You can make these `const`

File: src/jit/ee_il_dll.hpp [hunk 2]
```diff
@@ -109,6 +109,13 @@ inline var_types Compiler::eeGetArgType(CORINFO_ARG_LIST_HANDLE list, CORINFO_SI
     return JITtype2varType(strip(type));
 }
 
+/*****************************************************************************/
+inline CORINFO_CLASS_HANDLE Compiler::eeGetArgClass(CORINFO_ARG_LIST_HANDLE list, CORINFO_SIG_INFO* sig)
+{
+    CORINFO_CLASS_HANDLE argClass = info.compCompHnd->getArgClass(sig, list);
```
Comment (BruceForstall) [inline src/jit/ee_il_dll.hpp]: Seems odd that you would reverse the arg order.

File: src/jit/flowgraph.cpp [hunk 3]
```diff
@@ -13082,6 +13080,14 @@ bool flowList::setEdgeWeightMaxChecked(BasicBlock::weight_t newWeight, BasicBloc
     return result;
 }
 
+void flowList::setNewEdgeWeights(BasicBlock::weight_t newMinWeight, BasicBlock::weight_t newMaxWeight)
```
Comment (BruceForstall) [inline src/jit/flowgraph.cpp]: Needs function header

File: src/jit/block.h [hunk 4]
```diff
@@ -1289,13 +1299,14 @@ struct flowList
     //
     bool setEdgeWeightMinChecked(BasicBlock::weight_t newWeight, BasicBlock::weight_t slop, bool* wbUsedSlop);
     bool setEdgeWeightMaxChecked(BasicBlock::weight_t newWeight, BasicBlock::weight_t slop, bool* wbUsedSlop);
+    void setNewEdgeWeights(BasicBlock::weight_t newMinWeight, BasicBlock::weight_t newMaxWeight);
```
Comment (BruceForstall) [inline src/jit/block.h]: Could be just `setEdgeWeigths` -- no need for `New`.

**Comment by briansull** [comment]
@BruceForstall  @dotnet/jit-contrib PTAL

**Comment by BruceForstall** [comment]
I don't think you should add the JitOrderParser to the coreclr repo; I think you should put it in the jitutils repo instead. The jitOrderParser.cs file also needs a license header, and the filenames there should probably be all lower case to avoid future Linux issues.

**Comment by jashook** [comment]
> I don't think you should add the JitOrderParser to the coreclr repo; I think you should put it in the jitutils repo instead. The jitOrderParser.cs file also needs a license header, and the filenames there should probably be all lower case to avoid future Linux issues.

I agree

File: src/tools/JitOrderParser/jitOrderParser.cs [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        static void Main(string[] args)
+        {
+            int len = args.Length;
```
Comment (jashook) [inline src/tools/JitOrderParser/jitOrderParser.cs]: This seems to be missing usage documentation

File: src/jit/compiler.cpp [hunk 6]
```diff
@@ -5584,31 +5589,42 @@ void Compiler::compCompileFinish()
         {
             // clang-format off
             headerPrinted = true;
-            printf("         |  Profiled  | Exec-    |   Method has    |   calls   | Num |LclV |AProp| CSE |   Reg   |bytes | %3s code size | \n", Target::g_tgtCPUName);
-            printf(" mdToken |     |  RGN |    Count | EH | FRM | LOOP | NRM | IND | BBs | Cnt | Cnt | Cnt |  Alloc  |  IL  |   HOT |  COLD | method name \n");
-            printf("---------+-----+------+----------+----+-----+------+-----+-----+-----+-----+-----+-----+---------+------+-------+-------+-----------\n");
-            //      06001234 | PRF |  HOT |      219 | EH | ebp | LOOP |  15 |   6 |  12 |  17 |  12 |   8 |   28 p2 |  145 |   211 |   123 | System.Example(int)
+            printf("         |  Profiled   | Method   |   Method has    |   calls   | Num |LclV |AProp| CSE |   Perf  |bytes | %3s codesize| \n", Target::g_tgtCPUName);
```
Comment (jashook) [inline src/jit/compiler.cpp]: What perfscore gets output on platforms that are missing full support like arm*?

File: src/tools/JitOrderParser/JitOrderParser.csproj [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
Comment (jashook) [inline src/tools/JitOrderParser/JitOrderParser.csproj]: I think at this point we should target netcoreapp3.0

File: src/tools/JitOrderParser/JitOrderParser.sln [hunk 8]
```diff
@@ -0,0 +1,25 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (jashook) [inline src/tools/JitOrderParser/JitOrderParser.sln]: I think the general suggestion is to not check in sln files.

File: src/jit/compiler.cpp (hunk 6 above)
Comment (briansull) [reply src/jit/compiler.cpp]: The default PerfScore algorithm counts each instruction executed as 1 unit.

File: src/tools/JitOrderParser/JitOrderParser.sln (hunk 8 above)
Comment (briansull) [reply src/tools/JitOrderParser/JitOrderParser.sln]: OK

File: src/tools/JitOrderParser/JitOrderParser.csproj (hunk 7 above)
Comment (briansull) [reply src/tools/JitOrderParser/JitOrderParser.csproj]: OK

File: src/tools/JitOrderParser/jitOrderParser.cs (hunk 5 above)
Comment (briansull) [reply src/tools/JitOrderParser/jitOrderParser.cs]: This tool is still just a work in progress. I will move it over to jitutils and work on it some more.

**Comment by briansull** [comment]
@BruceForstall @jashook @dotnet/jit-contrib PTAL
I have addressed the feedback from Bruce and Jarret

With this change we will get unique hashes for all methods that have different signatures.

**Comment by AndyAyersMS** [comment]
Seems like the edge weight changes are unrelated to the rest and should be a separate PR. I'd hoped that we could get rid of the min/max weight scheme as it's just confusing. Why bother having two weights? Which one is more trustworthy?

I wonder if the hash changes resolve #1957? Do we now get unique hashes for IL stubs?

There are a number of places already where we call `getArgClass` directly. I never really understood the point of the wrappers for jit interface calls but it seems like we should be consistent and either always directly call or always call via a wrapper.

**Comment by briansull** [comment]
> I wonder if the hash changes resolve #1957? Do we now get unique hashes for IL stubs?

Yes, IL Stubs are uniquely identified with the new methodhash

**Comment by briansull** [comment]
I ran into an assert with edge weights when doing IBC training and optimization using a checked build.
This edge weight work was done to allow me to find and fix the bug.
I would like to get this work checked in, redesign is a separate issue here.

**Comment by briansull** [comment]
@AndyAyersMS if you really want I can checked these items in separately.

File: src/jit/morph.cpp [hunk 9]
```diff
@@ -14975,22 +14974,34 @@ bool Compiler::fgFoldConditional(BasicBlock* block)
...
                     // Now fix the weights of the edges out of 'bUpdated'
                     switch (bUpdated->bbJumpKind)
                     {
                         case BBJ_NONE:
-                            edge                  = fgGetPredForBlock(bUpdated->bbNext, bUpdated);
-                            edge->flEdgeWeightMax = bUpdated->bbWeight;
+                            edge         = fgGetPredForBlock(bUpdated->bbNext, bUpdated);
+                            newMaxWeight = bUpdated->bbWeight;
+                            newMinWeight = min(edge->edgeWeightMin(), newMaxWeight);
+                            edge->setEdgeWeights(newMinWeight, newMaxWeight);
```
Comment (briansull) [inline src/jit/morph.cpp]: These assignments of flEdgeWeightMax  could violate the requirements that  min-weight <= maxWeight

**Comment by AndyAyersMS** [comment]
> @AndyAyersMS if you really want I can checked these items in separately

I think it's a good practice not to mix unrelated (or seemingly unrelated) changes in one PR, but I'm not going to ask you to split this one up, as the rest of the changes are diagnostic.

We don't have a good mechanism for creating repros or regression tests for IBC related bugs. Seems like we ought to figure out how to do this.

**Comment by briansull** [comment]
> We don't have a good mechanism for creating repros or regression tests for IBC related bugs. Seems like we ought to figure out how to do this.

I agree that we should have a better way to create regression repros in this area.  
(Checking in the IBC data and running the IbcMerge tool are required steps)
I believe that their is some rework going on in the profile area,
when that comes online we should have the ability to added regression testing.

**Comment by briansull** [comment]
@BruceForstall  @jashook  @dotnet/jit-contrib PTAL

 This is ready for checkin

**Comment by BruceForstall** [comment]
> There are a number of places already where we call getArgClass directly. I never really understood the point of the wrappers for jit interface calls but it seems like we should be consistent and either always directly call or always call via a wrapper.

There's no need for a wrapper unless the wrapper does additional work. In many cases, we use wrappers so the altjit case can do something different.

**Pull request merged**